### PR TITLE
upgrade-controller target correct model

### PR DESCRIPTION
**Pull request by nammn**
##  Description of change

forces upgrade-controller to always use the controllermodel

### create user
- have bootstrapped env
```
juju add-user nam
juju register nam <token>
juju logout
juju login -u nam
```

### with rights
```
juju  grant nam superuser
juju upgrade-controller
```

### without rights
```
juju upgrade-controller
ERROR upgrade not possible missing permissions, current level "login", need "superuser"
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1850941

### CODE REVIEW

File: cmd/modelcmd/clientstore.go [hunk 1]
```diff
@@ -27,6 +28,12 @@ func (s QualifyingClientStore) QualifiedModelName(controllerName, modelName stri
 		return "", nil
 	}
 	if !jujuclient.IsQualifiedModelName(modelName) {
+		// If the name is not a qualified name but the modelName is still "controller"
+		// we can assume that we actually want to access the controller
+		// return the fully qualified controller name if we try to access the controller model
+		if modelName == bootstrap.ControllerModelName {
```
Comment (nammn) [inline cmd/modelcmd/clientstore.go]: We might want to discuss whether we want this to be here.
Maybe that would be too invasive

Comment (nammn) [reply cmd/modelcmd/clientstore.go]: @mitechie ^ It all depends on whether we want this behavior in other places as well (switch/show-model/offer...).

**Comment by mitechie** [comment]
No, we don't want it for everything. A user on the system is allowed to have a rick/controller model.

This is specific because when you say "upgrade-controller" that is specific as to where that happens an only users of superuser access can do that.

File: cmd/juju/commands/upgradecontroller.go [hunk 2]
```diff
@@ -262,19 +272,13 @@ func initCAASVersions(
...
 	jcmd := &upgradeJujuCommand{baseUpgradeCommand: baseUpgradeCommand{
 		upgradeMessage: "upgrade to this version by running\n    juju upgrade-controller",
 	}}
 	jcmd.SetClientStore(c.ClientStore())
 	wrapped := modelcmd.Wrap(jcmd)
-	controllerName, err := c.ControllerName()
-	if err != nil {
-		return errors.Trace(err)
-	}
-	fullControllerModelName := modelcmd.JoinModelName(controllerName,
```
Comment (manadart) [inline cmd/juju/commands/upgradecontroller.go]: I think you still need this bit, which explicitly prepends the controller name.

Comment (nammn) [reply cmd/juju/commands/upgradecontroller.go]: thanks, I've amended!

File: cmd/juju/commands/upgradecontroller.go [hunk 3]
```diff
@@ -262,19 +273,13 @@ func initCAASVersions(
...
-		bootstrap.ControllerModelName)
-	args := append(c.rawArgs, "-m", fullControllerModelName)
+	args := append(c.rawArgs, "-m", controllerModel)
```
Comment (mitechie) [inline cmd/juju/commands/upgradecontroller.go]: Do we have to have this? I don't think it makes sense for the user to provide what controller model is. Just upgrade-controller and leave it to Juju to find the right model name.

**Comment by nammn** [comment]
$$merge$$

**Comment by nammn** [comment]
$$merge$$

**Pull request merged**